compiler/src/dmd/templatesem.d

```diff
@@ -61,25 +61,142 @@ alias funcLeastAsSpecialized = dmd.funcsem.leastAsSpecialized;
+    }
+    return hash;
+}
+
+/************************************
+ * Computes hash of expression.
+ * Handles all Expression classes and MUST match their equals method,
+ * i.e. e1.equals(e2) implies expressionHash(e1) == expressionHash(e2).
+ */
+private size_t expressionHash(Expression e)
+{
+    import dmd.root.ctfloat : CTFloat;
+    import dmd.root.hash : calcHash, mixHash;
+
+    switch (e.op)
+    {
+    case EXP.int64:
+        return cast(size_t) e.isIntegerExp().getInteger();
+
+    case EXP.float64:
+        return CTFloat.hash(e.isRealExp().value);
+
+    case EXP.complex80:
+        auto ce = e.isComplexExp();
+        return mixHash(CTFloat.hash(ce.toReal), CTFloat.hash(ce.toImaginary));
+
+    case EXP.identifier:
+        return cast(size_t)cast(void*) e.isIdentifierExp().ident;
+
+    case EXP.null_:
+        return cast(size_t)cast(void*) e.isNullExp().type;
+
+    case EXP.string_:
+        return calcHash(e.isStringExp.peekData());
+
+    case EXP.tuple:
+    {
+        auto te = e.isTupleExp();
+        size_t hash = 0;
+        hash += te.e0 ? expressionHash(te.e0) : 0;
+        foreach (elem; *te.exps)
+            hash = mixHash(hash, expressionHash(elem));
+        return hash;
+    }
+
+    case EXP.arrayLiteral:
+    {
+        auto ae = e.isArrayLiteralExp();
+        size_t hash;
+        foreach (i; 0 .. ae.elements.length)
+            hash = mixHash(hash, expressionHash(ae[i]));
+        return hash;
+    }
+
+    case EXP.assocArrayLiteral:
+    {
+        auto ae = e.isAssocArrayLiteralExp();
+        size_t hash;
+        foreach (i; 0 .. ae.keys.length)
+            // reduction needs associative op as keys are unsorted (use XOR)
+            hash ^= mixHash(expressionHash((*ae.keys)[i]), expressionHash((*ae.values)[i]));
+        return hash;
+    }
+
+    case EXP.structLiteral:
+    {
+        auto se = e.isStructLiteralExp();
+        size_t hash;
+        foreach (elem; *se.elements)
+            hash = mixHash(hash, elem ? expressionHash(elem) : 0);
+        return hash;
+    }
+
+    case EXP.variable:
+        return cast(size_t)cast(void*) e.isVarExp().var;
+
+    case EXP.function_:
+        return cast(size_t)cast(void*) e.isFuncExp().fd;
+
+    default:
+        // no custom equals for this expression
+        //assert((&e.equals).funcptr is &Expression.equals);
+        // equals based on identity
+        return cast(size_t)cast(void*) e;
+    }
+}
+
 /************************************
  * This struct is needed for TemplateInstance to be the key in an associative array.
  * Fixing https://issues.dlang.org/show_bug.cgi?id=15813 would make it unnecessary.
  */
 struct TemplateInstanceBox
 {
     TemplateInstance ti;
+    size_t hash; // cached result of toHash()
 
     this(TemplateInstance ti)
     {
         this.ti = ti;
-        this.ti.toHash();
-        assert(this.ti.hash);
+        this.toHash();
+        assert(this.hash);
     }
 
-    size_t toHash() const @safe pure nothrow
```

Comment:
this needs to be `const`, I'd inline the hash function into the constructor above and leave this as is